### PR TITLE
manual compact: update shell tool

### PR DESCRIPTION
**Pull request by acelyc111**

### CODE REVIEW

File: scripts/pegasus_manual_compact.sh [hunk 1]
```diff
@@ -1,95 +1,261 @@
...
   echo "This tool is for manual compact specified table(app)."
-  echo "USAGE: $0 <cluster-meta-list> <app-name> [opts]"
-  echo "where opts including:"
-  echo "  ============================================================================="
-  echo "  | Name                        | ValueType                         | Default |"
-  echo "  |---------------------------------------------------------------------------|"
-  echo "  | target_level                | number in range of [1,num_levels] | -1      |"
-  echo "  | bottommost_level_compaction | skip or force                     | skip    |"
-  echo "  ============================================================================="
+  echo "USAGE: $0 -c cluster -a app-name [-t periodic|once] -g trigger-time [-d true|false] [-o k1=v1,k2=v2]"
```
Comment (LoveHeat) [inline scripts/pegasus_manual_compact.sh]: -o 参数没用了？

File: scripts/pegasus_manual_compact.sh [hunk 2]
```diff
@@ -1,95 +1,265 @@
...
-          left_time=$((sleeped * started_count / finish_count - sleeped))
+        query_log_file="/tmp/$UID.pegasus.query_compact.${app_id}"
+        echo "${query_cmd}" | ./run.sh shell --cluster ${cluster} &>${query_log_file}
+
+        queue_count=`grep 'recent enqueue at' ${query_log_file} | grep -v 'recent start at' | wc -l`
+        running_count=`grep 'recent start at' ${query_log_file} | wc -l`
+        not_finish_count=$((queue_count+running_count))
+        finish_count=`grep "last finish at" ${query_log_file} | grep -v "recent enqueue at" | grep -v "recent start at" | awk -F"[\[\]]" 'BEGIN{count=0}{if(length($2)==23 && $2>=$earliest_finish_time_ms){count++;}}END{print count}'`
+
+        if [ ${finish_count} -eq ${total_replica_count} ]; then
```
Comment (qinzuoyan) [inline scripts/pegasus_manual_compact.sh]: ${not_finish_count} -eq 0 and ${finish_count} -eq ${total_replica_count}

File: scripts/pegasus_manual_compact.sh [hunk 3]
```diff
@@ -1,95 +1,265 @@
...
+function usage()
+{
+    echo "This tool is for manual compact specified table(app)."
+    echo "USAGE: $0 -c cluster -a app-name [-t periodic|once] -g trigger-time [-d true|false] [...]"
+    echo "Options:"
+    echo "  -h|--help"
+    echo "  -c|--cluster          cluster meta server list, default is \"127.0.0.1:34601,127.0.0.1:34602\""
+    echo "  -a|--app_name         manual compact target table(app) name"
+    echo "  -t|--type             manual compact type, should be periodic or once, default is once"
+    echo "  -g|--trigger_time     manual compact trigger time"
```
Comment (qinzuoyan) [inline scripts/pegasus_manual_compact.sh]: --trigger_time 和 --disable_periodic 只对 -t periodic 有效

File: scripts/pegasus_manual_compact.sh [hunk 4]
```diff
@@ -11,14 +11,16 @@ function usage()
...
-    echo "                        unix timestamp format for once type, e.g. \"1514736000\" for Jan. 1, 00:00:00 CST 2018"
     echo "  -d|--disable_periodic whether to disable periodic manual compact, default is false which is not disable"
     echo "  --target_level        number in range of [1,num_levels], default is -1"
     echo "  --bottommost_level_compaction     skip or force, default is skip"
     echo "                        more details: https://github.com/facebook/rocksdb/wiki/Manual-Compaction"
     echo
     echo "for example:"
-    echo "$0 127.0.0.1:34601,127.0.0.1:34602 -t periodic -g 3:00,21:00 -a temp --target_level 2 --bottommost_level_compaction force"
+    echo "  once type manual compact with default options:"
+    echo "    $0 127.0.0.1:34601,127.0.0.1:34602 -a temp"
```
Comment (qinzuoyan) [inline scripts/pegasus_manual_compact.sh]: $0 -c 127.0.0.1:34601,127.0.0.1:34602

File: scripts/pegasus_manual_compact.sh [hunk 5]
```diff
@@ -11,14 +11,16 @@ function usage()
...
+    echo "  periodic type manual compact with specified options:"
+    echo "    $0 127.0.0.1:34601,127.0.0.1:34602 -a temp -t periodic -g 3:00,21:00 --target_level 2 --bottommost_level_compaction force"
```
Comment (qinzuoyan) [inline scripts/pegasus_manual_compact.sh]: $0 -c 127.0.0.1:34601,127.0.0.1:34602

File: scripts/pegasus_manual_compact.sh [hunk 6]
```diff
@@ -1,95 +1,266 @@
...
-          left_time=$((sleeped * started_count / finish_count - sleeped))
+        query_log_file="/tmp/$UID.pegasus.query_compact.${app_id}"
+        echo "${query_cmd}" | ./run.sh shell --cluster ${cluster} &>${query_log_file}
+
+        queue_count=`grep 'recent enqueue at' ${query_log_file} | grep -v 'recent start at' | wc -l`
+        running_count=`grep 'recent start at' ${query_log_file} | wc -l`
+        not_finish_count=$((queue_count+running_count))
+        finish_count=`grep "last finish at" ${query_log_file} | grep -v "recent enqueue at" | grep -v "recent start at" | awk -F"[\[\]]" -v date="$earliest_finish_time_ms" 'BEGIN{count=0}{if(length($2)==23 && $2>=date){count++;}}END{print count}'`
+
+        if [ ${finish_count} -eq ${total_replica_count} ]; then
```
Comment (qinzuoyan) [inline scripts/pegasus_manual_compact.sh]: ${not_finish_count} -eq 0 and ${finish_count} -eq ${total_replica_count}

**Pull request merged**